### PR TITLE
[gdal_tindex] Use argparser

### PR DESCRIPTION
**Pull request by elpaso**

### CODE REVIEW

File: apps/argparse/argparse.hpp [hunk 1]
```diff
@@ -741,6 +742,20 @@ class Argument {
     return *this;
   }
 
+  auto &store_into(std::set<std::string> &var) {
```
Comment (rouault) [inline apps/argparse/argparse.hpp:745]: yes, please try to upstream that. See https://github.com/p-ranav/argparse/pull/343 for a potential similar contribution. The tests have to be manually run with "test/tests" from the CMake build directory

File: apps/gdaltindex_lib.cpp [hunk 2]
```diff
@@ -165,6 +166,231 @@ static bool PatternMatch(const char *input, const char *pattern)
...
+    argParser->add_argument("-filename_filter")
+        .metavar("<val>")
+        .append()
+        .store_into(psOptions->oSetFilenameFilters)
+        .help(_("Pattern that the filenames contained in directories pointed "
+                "by <file_or_dir> should follow. '*' is a wildcard character "
+                "that matches any number of any characters including none. '?' "
+                "is a wildcard character that matches a single character. "
+                "Comparisons are done in a case insensitive way. Several "
+                "filters may be specified.."));
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:207]: a bit too long:
```suggestion
        .help(_("Pattern that the filenames contained in directories pointed "
                "by <file_or_dir> should follow."));
```

File: apps/gdaltindex_lib.cpp [hunk 3]
```diff
@@ -165,6 +166,231 @@ static bool PatternMatch(const char *input, const char *pattern)
...
+
+    argParser->add_argument("-min_pixel_size")
+        .metavar("<val>")
+        .store_into(psOptions->dfMinPixelSize)
+        .help(_("Minimum pixel size in term of geospatial extent per pixel "
+                "(resolution) that a raster should have to be selected. The "
+                "pixel size is evaluated after reprojection of its extent to "
+                "the target SRS defined by -t_srs."));
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:215]:
```suggestion
        .help(_("Minimum pixel size in term of geospatial extent per pixel "
                "(resolution) that a raster should have to be selected."));
```

File: apps/gdaltindex_lib.cpp [hunk 4]
```diff
@@ -165,6 +166,231 @@ static bool PatternMatch(const char *input, const char *pattern)
...
+
+    argParser->add_argument("-max_pixel_size")
+        .metavar("<val>")
+        .store_into(psOptions->dfMaxPixelSize)
+        .help(_("Maximum pixel size in term of geospatial extent per pixel "
+                "(resolution) that a raster should have to be selected. The "
+                "pixel size is evaluated after reprojection of its extent to "
+                "the target SRS defined by -t_srs."));
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:223]:
```suggestion
        .help(_("Maximum pixel size in term of geospatial extent per pixel "
                "(resolution) that a raster should have to be selected."));
```

File: apps/gdaltindex_lib.cpp [hunk 5]
```diff
@@ -165,6 +166,231 @@ static bool PatternMatch(const char *input, const char *pattern)
...
+
+    argParser->add_argument("-skip_different_projection")
+        .flag()
+        .store_into(psOptions->bSkipDifferentProjection)
+        .help(_(
+            "Only files with the same projection as files already inserted in "
+            "the tile index will be inserted (unless -t_srs is specified)."));
+
+    argParser->add_argument("-t_srs")
+        .metavar("<target_srs>")
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:246]: We use srs_def for similar instances:
```suggestion
        .metavar("<srs_def>")
```

File: apps/gdaltindex_lib.cpp [hunk 6]
```diff
@@ -165,6 +166,231 @@ static bool PatternMatch(const char *input, const char *pattern)
...
+        .help(_("Name of the layer in the tile index file."));
+
+    argParser->add_argument("-lco")
+        .metavar("<KEY>=<VALUE>")
+        .append()
+        .action(
+            [psOptions](const std::string &s)
+            {
+                if (psOptions)
+                    psOptions->aosLCO.AddString(s.c_str());
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:286]: psOptions can't be null:
```suggestion
                psOptions->aosLCO.AddString(s.c_str());
```

File: apps/gdaltindex_lib.cpp [hunk 7]
```diff
@@ -165,6 +166,231 @@ static bool PatternMatch(const char *input, const char *pattern)
...
+                    psOptions->eSrcSRSFormat = FORMAT_AUTO;
+            })
+        .help(_("Format of the source SRS to store in the tile index file."));
+
+    argParser->add_argument("-lyr_name")
+        .metavar("<name>")
+        .store_into(psOptions->osIndexLayerName)
+        .help(_("Name of the layer in the tile index file."));
+
+    argParser->add_argument("-lco")
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:279]: could be worth to have an helper for that in GDALArgumentParser. That's used in several utilities

File: apps/gdaltindex_lib.cpp [hunk 8]
```diff
@@ -165,6 +166,231 @@ static bool PatternMatch(const char *input, const char *pattern)
...
+        .help(_("Color interpretation of of the tiles of the tile index: red, "
+                "green, blue, alpha, gray, undefined."));
+
+    argParser->add_argument("-mask")
+        .flag()
+        .store_into(psOptions->bMaskBand)
+        .help(_("Add a mask band to the tiles of the tile index."));
+
+    argParser->add_argument("-mo")
+        .metavar("<name=value>")
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:336]:
```suggestion
        .metavar("<name>=<value>")
```

File: apps/gdaltindex_lib.cpp [hunk 9]
```diff
@@ -1167,218 +1366,123 @@ GDALTileIndexOptionsNew(char **papszArgv,
...
-            else if (EQUAL(pszType, "Integer64"))
-                md.eType = OFTInteger64;
-            else if (EQUAL(pszType, "Real"))
-                md.eType = OFTReal;
-            else if (EQUAL(pszType, "Date"))
-                md.eType = OFTDate;
-            else if (EQUAL(pszType, "DateTime"))
-                md.eType = OFTDateTime;
-            else
+            if (fetchMd->size() % 3)
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:1408]: is this check really needed ? argparse constraints should have validated that already, shouldn't they ? Or perhaps make it a CPLAssert() ?

File: apps/gdaltindex_lib.cpp [hunk 10]
```diff
@@ -1167,218 +1366,123 @@ GDALTileIndexOptionsNew(char **papszArgv,
...
+                }
+                else if (typeName == "Real")
+                {
+                    type = OFTReal;
+                }
+                else if (typeName == "Date")
+                {
+                    type = OFTDate;
+                }
+                // "Time" is not documented: is it really unsupported?
```
Comment (rouault) [inline apps/gdaltindex_lib.cpp:1442]: in practice for an image catalog, a time without date is of little use

File: apps/gdaltindex_lib.cpp (hunk 9 above)
Comment (elpaso) [reply apps/gdaltindex_lib.cpp:1408]: no, it's a leftover for when I thought it supported multiple 3*n arguments. I'll remove it.

File: apps/argparse/argparse.hpp (hunk 1 above)
Comment (elpaso) [reply apps/argparse/argparse.hpp:745]: Done: https://github.com/p-ranav/argparse/pull/348

File: apps/gdaltindex_lib.cpp (hunk 9 above)
Comment (elpaso) [reply apps/gdaltindex_lib.cpp]: done (CPLAssert)

File: apps/gdaltindex_lib.cpp (hunk 10 above)
Comment (elpaso) [reply apps/gdaltindex_lib.cpp]: removed

File: apps/gdaltindex_lib.cpp (hunk 7 above)
Comment (elpaso) [reply apps/gdaltindex_lib.cpp]: done

**Comment by rouault** [comment]
Do you mind reworking the commit history in 4 separate commits?:
- change to argparse.hpp
- addition of the new method in gdalargumentparser.h/cpp
- use of this new method in the already converted utilities (I wouldn't mind if it is merged in the previous commit)
- changes to gdaltindex_bin/_lib

**Comment by elpaso** [comment]
> Do you mind reworking the commit history in 4 separate commits?:
> 
>     * change to argparse.hpp
> 
>     * addition of the new method in gdalargumentparser.h/cpp
> 
>     * use of this new method in the already converted utilities (I wouldn't mind if it is merged in the previous commit)
> 
>     * changes to gdaltindex_bin/_lib

done

**Pull request merged**